src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/markduplicates/MarkDuplicatesFromShardsDataflowTransform.java

```diff
@@ -0,0 +1,165 @@
+                    shipping = Stopwatch.createUnstarted();
+                }
+
+                @Override
+                public void finishBundle(Context c) throws Exception {
+                    super.finishBundle(c);
+                    log.info("MarkDuplicatesFromShardsDataflow bundle worked: "+processing.elapsed(TimeUnit.MILLISECONDS)+" ms. Shipped: "+shipping.elapsed(TimeUnit.MILLISECONDS)+" ms.");
+                }
+
+                // Convert from PCollection to LocalCollection, run the computation there, convert back to PCollection.
+                @Override
+                public void processElement(ProcessContext c) throws Exception {
+                    processing.start();
+                    final SAMFileHeader h = c.sideInput(header);
+                    final OpticalDuplicateFinder f = c.sideInput(finder);
+                    LocalCollection<GATKRead> answer = localApply(h, f, c.element().getValue());
+                    processing.stop();
+                    shipping.start();
+                    for (GATKRead r : answer.iterable()) {
+                        c.output(r);
+                    }
+                    shipping.stop();
+                }
+            }));
+    }
+
+    public static LocalCollection<GATKRead> localApply(final SAMFileHeader header, final OpticalDuplicateFinder finder, Iterable<GATKRead> readsIterable) {
+
+        // read input, make our own copy.
+        LocalCollection<GATKRead> reads = LocalCollection.of(readsIterable)
+            .map(MarkDuplicatesUtils::markNonDuplicate);
+
+        if (reads.size()==0) return reads;
+
+        // we specify a minimum size to guarantee we always get 2 elements in the array, even if e.g. all the reads are primary.
+        ArrayList<LocalCollection<GATKRead>> readsPartitioned = reads.partitionBy(MarkDuplicates::primaryNonPrimaryAlignment, 2);
+        LocalCollection<GATKRead> fragments = readsPartitioned.get(MarkDuplicates.ReadsPartition.PRIMARY.ordinal());
+        LocalCollection<GATKRead> fragmentsTransformed = transformFragments(header, fragments);
+
+        LocalCollection<GATKRead> pairs = readsPartitioned.get(MarkDuplicates.ReadsPartition.PRIMARY.ordinal());
+        LocalCollection<GATKRead> pairsTransformed = transformReads(header, finder, pairs);
+
+        LocalCollection<GATKRead> not_primary = readsPartitioned.get(MarkDuplicates.ReadsPartition.NOT_PRIMARY.ordinal());
+
+        LocalCollection<GATKRead> ret = LocalCollection.union(fragmentsTransformed, pairsTransformed, not_primary)
+            .map(MarkDuplicatesFromShardsDataflowTransform::removeHeader);
+
+        return ret;
+    }
+
+
+    private static LocalCollection<GATKRead> transformFragments(final SAMFileHeader header, LocalCollection<GATKRead> fragments) {
+        return fragments
+            // group by key (genomic location etc)
+            .groupBy((read) -> MarkDuplicatesUtils.keyForFragment(header, read))
```

Comment:
I think you have shard boundary issues in this implementation of MarkDupes: since MD keys the reads using (among other things) the stranded unclipped start (in `ReadsKey.keyForFragment()`), and the sharding is done using simple start/stop positions, I think it could easily happen that reads sharing the same key according to `MarkDuplicatesUtils.keyForFragment()` could end up in different shards due to the presence of clipped bases. In these cases, the tool will fail to find and mark all duplicates. This was not possible in the un-optimized version, since we explicitly grouped by the key returned from `MarkDuplicatesUtils.keyForFragment()`